File: day_12/task_1.rb
DIRECTIONS = %w(N W S E)
ROTATIONS = %w(L R)
instructions = File.read("input.txt").split("\n")

def next_position(position, direction, distance)
  case direction
  when "N"
    [position[0], position[1] + distance]
  when "S"
    [position[0], position[1] - distance]
  when "E"
    [position[0] + distance, position[1]]
  when "W"
    [position[0] - distance, position[1]]
  end
end

def next_direction(direction, rotation_direction, rotation)
  if rotation_direction == "R"
    DIRECTIONS[DIRECTIONS.index(direction) - (rotation / 90) % 4]
  elsif rotation_direction == "L"
    DIRECTIONS[DIRECTIONS.index(direction) - DIRECTIONS.length + (rotation / 90) % 4]
  end
end

ship_direction = "E"
ship_position = [0, 0]

instructions.each do |instruction|
  command, value = instruction[0], instruction[1..-1].to_i

  if DIRECTIONS.include?(command)
    ship_position = next_position(ship_position, command, value)
  elsif ROTATIONS.include?(command)
    ship_direction = next_direction(ship_direction, command, value)
  elsif command == "F"
    ship_position = next_position(ship_position, ship_direction, value)
  end
end

p ship_position.map(&:abs).sum

File: 2020/day_12/task_1.py
DIRECTIONS = ["N", "W", "S", "E"]
ROTATIONS = ["L", "R"]

def next_position(position, direction, distance):
  if direction == "N":
    return [position[0], position[1] + distance]
  elif direction == "S":
    return [position[0], position[1] - distance]
  elif direction == "E":
    return [position[0] + distance, position[1]]
  elif direction == "W":
    return [position[0] - distance, position[1]]

def next_direction(direction, rotation_direction, rotation):
  if rotation_direction == "R":
    return DIRECTIONS[DIRECTIONS.index(direction) - int((rotation / 90) % 4)]
  elif rotation_direction == "L":
    return DIRECTIONS[DIRECTIONS.index(direction) - len(DIRECTIONS) + int((rotation / 90) % 4)]

ship_direction = "E"
ship_position = [0, 0]

file = open("input.txt", "r")

for instruction in file.readlines():
  instruction = instruction.strip("\n")
  command, value = instruction[0], int(instruction[1:])

  if command in DIRECTIONS:
    ship_position = next_position(ship_position, command, value)
  elif command in ROTATIONS:
    ship_direction = next_direction(ship_direction, command, value)
  elif command == "F":
    ship_position = next_position(ship_position, ship_direction, value)

file.close()

print(sum(map(abs, ship_position)))
